source/jaster/cli/core.d

```diff
@@ -117,6 +117,371 @@ ServiceInfo[] addCommandLineInterfaceService(ref ServiceInfo[] services)
+        cast(ArgIsOptional)arg.isNullable,
+        arg.uda.name
+        );
+    }
+
+    builder.commandName = UDA.pattern;
+    builder.description = UDA.description;
+
+    return builder;
+}
+
+private CommandCompleteFunc createCommandCompleteFunc(alias T)(CommandArguments!T commandArgs)
+{
+    import std.algorithm : filter, map, startsWith, splitter, canFind;
+    import std.exception : assumeUnique;
+
+    return (string[] before, string current, string[] after, ref char[] output)
+    {
+        // Check if there's been a null ("--") or '-' ("---"), and if there has, don't bother with completion.
+        // Because anything past that is of course, the raw arg list.
+        if(before.canFind(null) || before.canFind("-"))
+            return;
+
+        // See if the previous value was a non-boolean argument.
+        const justBefore               = ArgPullParser(before[$-1..$]).front;
+        auto  justBeforeNamedArgResult = commandArgs.namedArgs.filter!(a => matchSpacelessPattern(a.uda.pattern, justBefore.value));
+        if((justBefore.type == ArgTokenType.LongHandArgument || justBefore.type == ArgTokenType.ShortHandArgument)
+        && (!justBeforeNamedArgResult.empty && !justBeforeNamedArgResult.front.isBool))
+        {
+            // TODO: In the future, add support for specifying values to a parameter, either static and/or dynamically.
+            return;
+        }
+
+        // Otherwise, we either need to autocomplete an argument's name, or something else that's predefined.
+
+        string[] names;
+        names.reserve(commandArgs.namedArgs.length * 2);
+
+        foreach(arg; commandArgs.namedArgs)
+        {
+            foreach(pattern; arg.uda.pattern.byPatternNames)
+            {
+                // Reminder: Confusingly for this use case, arguments don't have their leading dashes in the before and after arrays.
+                if(before.canFind(pattern) || after.canFind(pattern))
+                    continue;
+
+                names ~= pattern;
+            }
+        }
+
+        foreach(name; names.filter!(n => n.startsWith(current)))
+        {
+            output ~= (name.length == 1) ? "-" : "--";
+            output ~= name;
+            output ~= ' ';
+        }
+    };
+}
+
+private CommandExecuteFunc createCommandExecuteFunc(alias T)(CommandArguments!T commandArgs)
+{
+    import std.format    : format;
+    import std.algorithm : filter, map;
+    import std.exception : enforce, collectException;
+
+    // This is expecting the parser to have already read in the command's name, leaving only the args.
+    return (ArgPullParser parser, ref string executionError, scope ref ServiceScope services, HelpTextBuilderSimple helpText)
+    {
+        if(containsHelpArgument(parser))
+        {
+            import std.stdio : writeln;
+            writeln(helpText.toString());
+            return 0;
+        }
+
+        // Cross-stage state.
+        T                     commandInstance;
```

Comment:
Can probably remove the alignment spacing here, now that those other two variables were moved.

This is just a nitpicky one, it can be left as-is.